python/cog/types.py

```diff
@@ -198,15 +203,31 @@ def __repr__(self) -> str:
 
 
 def get_filename(url: str) -> str:
-    parsed_url = urlparse(url)
+    parsed_url = urllib.parse.urlparse(url)
+
     if parsed_url.scheme == "data":
-        header, _ = parsed_url.path.split(",", 1)
-        mime_type, _ = header.split(";", 1)
+        resp = urllib.request.urlopen(url)
+        mime_type = resp.headers.get_content_type()
         extension = mimetypes.guess_extension(mime_type)
         if extension is None:
             return "file"
         return "file" + extension
-    return os.path.basename(parsed_url.path)
+
+    basename = os.path.basename(parsed_url.path)
+    basename = urllib.parse.unquote_plus(basename)
+
+    # If the filename is too long, we truncate it (appending '~' to denote the
+    # truncation) while preserving the file extension.
+    # - truncate it
+    # - append a tilde
+    # - preserve the file extension
+    if _len_bytes(basename) > FILENAME_MAX_LENGTH:
+        basename = _truncate_filename_bytes(basename, length=FILENAME_MAX_LENGTH)
+
+    for c in FILENAME_ILLEGAL_CHARS:
+        basename = basename.replace(c, "_")
```

Comment:
no strong preference on this, but nicer to replace multiple illegal chars in a row with a single `_`?

debugging with multiple _ might be complicated to count - but will probably never actually happen - so ignore me?

    pattern = "[" + re.escape(''.join(FILENAME_ILLEGAL_CHARS)) + "]+"
    basename = re.sub(pattern, "_", basename)

Reply:
We could do that, but I'm not sure there's a meaningful user experience improvement here (how many people have file names with multiple NULs or `/`s in a row?) and it would make the code a little harder to grok.

Reply:
agreed